bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+            TypeError: callable returns non-numeric (int, float) stepsize
+            TypeError: invalid type for stepsize
+
+        Returns:
+            list[float]: leapfrog stepsizes for self._num_proposal proposals
+        """
+        if type(stepsize) is int or type(stepsize) is float:
+            if not stepsize > 0:
+                raise ValueError(
+                    f"stepsize must be positive, but found stepsize of {stepsize}"
+                )
+            return [float(stepsize)] * self._num_proposals
+
+        elif type(stepsize) is list:
+            if len(stepsize) != self._num_proposals:
+                raise ValueError(
+                    f"list of stepsize must be of length {self._num_proposals}, but found {len(stepsize)} stepsizes"
+                )
+            for s in stepsize:
+                if not s > 0:
+                    raise ValueError(
+                        f"stepsize must be positive, but found stepsize of {s}"
+                    )
+                if type(s) is not int and type(s) is not float:
+                    raise TypeError(
+                        f"list of stepsizes must contain integers or floats, but found {type(s)} stepsize"
+                    )
+            return [float(s) for s in stepsize]
+
+        elif isinstance(stepsize, Callable):  # isinstance() allows for lambda functions
+            stepsize = [stepsize(k) for k in range(self._num_proposals)]
+            for s in stepsize:
+                if not s > 0:
+                    raise ValueError(
+                        f"stepsize must be positive, but found stepsize of {s}"
+                    )
+                if type(s) is not int and type(s) is not float:
+                    raise TypeError(
+                        f"callable must return integer or float stepsize, but found {type(s)} steps"
+                    )
+            return [float(s) for s in stepsize]
+
+        else:
+            raise TypeError("stepsize must be None, int, float, list, or callable")
+
+    def _init_steps(
+        self,
+        steps: Union[int, list[int], Callable[[int], int]],
+    ) -> list[int | float]:
+        """List with number of leapfrog steps used to generate each proposed sample.
+
+        Each proposed sample (theta_prop, rho_prop) is generated by taking multiple
+        leapfrog steps with a a fixed stepsize. Each distinct proposal, however, may
+        have a different number of steps. This function initializes the number of steps
+        for self._num_proposal such proposals.
+
+        This function accepts, among other valid inputs, a callable stepsize function.
+        This callable should accept the (integer) proposal number and return the
+        number of leapfrog steps for that proposal.
+
+        Intuitively, DrGhmcDiag reduces the stepsize in subsequent proposals if the
+        previous proposal was rejected for better resolution of high-curvature
+        state space regions.
```

Comment:
Should I add a warning if the leapfrog stepsizes are not monotonically decreasing?

Reply:
If that's a requirement for the underlying algorithm, and something that can be checked without introducing the need to carry a lot of state from one function call to another, then it sounds to me like a good idea. (In general, if execution is violating an algorithm's invariant or requirement, that's something the user should be made aware of, unless there's a really good reason not to.)

Reply:
It's not a requirement to reduce step size at each iteration.  The first two applications people did left it the same and continued from where the first proposal left off, for example.  I don't think this implementation will be general enough to do that since it's always re-proposing from the same starting state.

I would say it reduces the step size for subsequent proposals because it leads to a higher acceptance probability if the first proposal was rejected because the Hamiltonian diverged (i.e., was not preserved by the step).  The underlying problem is trying to follow curvature using only step size plus gradients.

Reply:
So should increasing stepsizes trigger an error?

Reply:
It could, but my feeling is to just leave the generality as it doesn't break the algorithm.

Reply:
(And, if the algorithm doesn't actually require XYZ, the comments probably shouldn't say that it does.)

Reply:
Let's make that "definitely" :-).  You can suggest that's the usual usage, but don't say it's required if it's not.  I'm generally suggesting not requiring it because it's a simpler function with less doc and less edge-case testing.  I am OK if you want to make it require a monotonic downward set of step sizes, but it should allow them to stay the same at least.